rtl/eclrtl/rtldynfield.cpp

```diff
@@ -510,3 +510,481 @@ extern ECLRTL_API int getFieldNum(const char *fieldName, IOutputMetaData &  meta
+            }
+        }
+        if (!canTranslate())
+            DBGLOG("Translation is NOT possible");
+        else if (needsTranslate())
+            DBGLOG("Translation is possible");
+        else
+            DBGLOG("Translation is not necessary");
+    }
+    size32_t translate(ARowBuilder &builder, size32_t offset, const byte *sourceRec) const
+    {
+        unsigned numOffsets = sourceRecInfo.getNumVarFields() + 1;
+        size_t variableOffsets[numOffsets];
+        RtlRow sourceRow(sourceRecInfo, sourceRec, numOffsets, variableOffsets);
+        for (int idx = 0; idx < destRecInfo.getNumFields(); idx++)
+        {
+            //printf("before idx %d offset %d\n", idx, offset);
+            const RtlFieldInfo *field = destRecInfo.queryField(idx);
+            const RtlTypeInfo *type = field->type;
+            const MatchInfo &match = matchInfo[idx];
+            if (match.matchType == match_none || match.matchType==match_fail)
+                offset = type->buildNull(builder, offset, field);
+            else
+            {
+                unsigned matchField = match.matchIdx;
+                const RtlTypeInfo *sourceType = sourceRecInfo.queryType(matchField);
+                size_t sourceOffset = sourceRow.getOffset(matchField);
+                const byte *source = sourceRec + sourceOffset;
+                size_t copySize = sourceRow.getOffset(matchField+1) - sourceOffset;
+                switch (match.matchType)
+                {
+                case match_perfect:
+                {
+                    // Look ahead for other perfect matches and combine the copies
+                    while (idx < destRecInfo.getNumFields()-1)
+                    {
+                        const MatchInfo &nextMatch = matchInfo[idx+1];
+                        if (nextMatch.matchType == match_perfect && nextMatch.matchIdx == matchField+1)
+                        {
+                            idx++;
+                            matchField++;
+                        }
+                        else
+                            break;
+                    }
+                    size_t copySize = sourceRow.getOffset(matchField+1) - sourceOffset;
+                    builder.ensureCapacity(offset+copySize, field->name);
+                    memcpy(builder.getSelf()+offset, source, copySize);
+                    offset += copySize;
+                    break;
+                }
+                case match_truncate:
+                {
+                    assert(type->isFixedSize());
+                    size32_t copySize = type->getMinSize();
+                    builder.ensureCapacity(offset+copySize, field->name);
+                    memcpy(builder.getSelf()+offset, source, copySize);
+                    offset += copySize;
+                    break;
+                }
+                case match_extend:
+                {
+                    size32_t destSize = type->size(source, sourceRec);
```

Comment:
I could put in the assert (it should never fire...)

I use the field for truncate because the amount to copy is dependent on the target field. Here the amount to memcpy is dependent on the source field (and the amount to pad is dependent on the target field). I don't think they are quite as similar as might first appear.

I could probably precalculate the sizes though, since restricted to fixed size.